### PR TITLE
[pkg/ottl] Add private interface for path and key

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:**
This is the first PR working towards https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22744.

It adds 2 new interfaces, `path` and `key`, that will be given to contexts to use when interpreting the paths/keys passed from the grammar. Eventually we'll be able to use the `isComplete` functions during parsing to ensure that all parts of the path were used by the context.

By the end of that issue we'll have removed all public API's of OTTL's grammar, which is going to cause massive breaking changes to all our contexts and huge PRs.  To help decrease the amount of changes per PR these new interfaces are not yet exported and not yet implemented.  To see them partially in use checkout https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29703.

**Link to tracking Issue:**

Related to https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22744

**Testing:**

Added unit tests

**Documentation:** <Describe the documentation added.>

I'll add godoc comments once these things are made public

### CODE REVIEW

File: pkg/ottl/functions.go [hunk 1]
```diff
@@ -18,6 +18,125 @@ type EnumParser func(*EnumSymbol) (*Enum, error)
...
+	return p.nextPath
+}
+
+func (p *basePath) Key() key {
+	return p.key
+}
+
+func (p *basePath) isComplete() error {
+	if !p.fetched {
+		return fmt.Errorf("path section '%v' was not fetched", p.name)
```
[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply pkg/ottl/functions.go]: Maybe.  We'll end up using this whenever we're done calling `parsePath`.  We could either call `isComplete` and then, if there is an error, build a user friendly error or just return the error.

Just returning seems better so it is probably worth making this error nicer.

[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply pkg/ottl/functions.go]: Updated

File: pkg/ottl/functions_test.go [hunk 2]
```diff
@@ -2093,3 +2093,184 @@ func defaultFunctionsForTests() map[string]Factory[any] {
...
+
+func Test_basePath_Key(t *testing.T) {
+	k := &baseKey{}
+	bp := basePath{
+		key: k,
+	}
+	assert.Equal(t, k, bp.Key())
+}
+
+func Test_basePath_isComplete(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply pkg/ottl/functions_test.go:2121]: Good idea.  This caught that baseKey's `Next()` implementation wasn't updating the fetched flag.

[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply pkg/ottl/functions_test.go:2121]: @evan-bradley updated

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**